platform/web/request.go

```diff
@@ -58,78 +80,38 @@ func DecodeWithJSONSchema(request *http.Request, model interface{}, filePath str
 
 	filePath = path.Join(path.Dir(rootFile), filePath)
 
-	jsonSchema := gojsonschema.NewReferenceLoader("file://" + filePath)
+	jsonSchema, err := os.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("%w", NewErrInvalidJSONSchemaFilePath())
+	}
 
-	if err := validateRequestPayload(request, model, options, jsonSchema); err != nil {
+	if err := DecodeWithEmbeddedJSONSchema(request, model, jsonSchema, options); err != nil {
 		return err
 	}
 	return nil
 }
 
 // DecodeWithEmbeddedJSONSchema takes json schema and a http request
 // And returns an error when the request's payload does not match the JSON schema
-func DecodeWithEmbeddedJSONSchema(request *http.Request, model interface{}, jsonSchemaBytes []byte, options DecoderOptions) error {
-	jsonSchema := gojsonschema.NewBytesLoader(jsonSchemaBytes)
-
-	if err := validateRequestPayload(request, model, options, jsonSchema); err != nil {
-		return err
-	}
-	return nil
-}
-
-func validateRequestPayload(request *http.Request, model interface{}, options DecoderOptions, jsonSchema gojsonschema.JSONLoader) error {
-	body, _ := io.ReadAll(request.Body)
-	request.Body = io.NopCloser(bytes.NewBuffer(body))
-	payload := gojsonschema.NewBytesLoader(body)
-
-	result, err := gojsonschema.Validate(jsonSchema, payload)
+func DecodeWithEmbeddedJSONSchema(request *http.Request, model interface{}, jsonSchema []byte, options DecoderOptions) error {
+	body, err := io.ReadAll(request.Body)
 	if err != nil {
-		return fmt.Errorf("%w: %v", NewErrBadRequestResponse(nil), err)
-	}
-
-	if !result.Valid() {
-		additionalPropertyErrors := make(ErrorDetails)
-		requiredPropertyErrors := make(ErrorDetails)
-		requestContentErrorDetails := make(ErrorDetails)
-
-		for _, desc := range result.Errors() {
-			propertyName := desc.Field()
-			if propertyName == "(root)" {
-				details, ok := desc.Details()["property"]
-				if ok {
-					propertyName = fmt.Sprintf("%v", details)
-				}
-			}
-
-			switch desc.Type() {
-			case "additional_property_not_allowed":
-				additionalPropertyErrors[propertyName] = desc.Description()
-			case "required":
-				requiredPropertyErrors[propertyName] = desc.Description()
-			default:
-				requestContentErrorDetails[propertyName] = desc.Description()
-			}
-		}
-
-		if len(additionalPropertyErrors) > 0 {
-			errDetails := additionalPropertyErrors
-			for property, errMessage := range requestContentErrorDetails {
-				errDetails[property] = errMessage
-			}
-
-			return fmt.Errorf("the request body contains unknown keys: %w", NewErrBadRequestResponse(errDetails))
+		if err.Error() == "http: request body too large" {
+			return fmt.Errorf("valid request payload failed: %w", ErrRequestBodyTooLarge)
 		}
+		return fmt.Errorf("could not read request body: %v", err)
+	}
+	request.Body = io.NopCloser(bytes.NewBuffer(body))
```

Comment:
wow this looks super dangerous and I am not even sure what exactly you are trying to achieve

Reply:
We had it before in `validateRequestPayload()` actually and I had a hard time understand it then when Bert added it 😅 , but it's actually not dangerous.

Since we are using `decode(r *http.Request, val interface{}, options DecoderOptions)` we're calling after JSON schema validation needs the request body that we've already read to validate against the JSON schema, we need to recreate it, so we make a new buffer with the body.

Then  `io.NopCloser` just adds a `Close()` method so that we go from a buffer that only implements `io.Reader` to something that implements `io.ReadCloser` which is the expected type for a request body.

We could just reuse the body in the `[]byte` form we already have at this point, but that means we would have to use `json.Unmarshal` and not `json.Decoder`.

Reply:
What I don't get is why we would reassign something in the request body. What is the need for that ?

Reply:
Well we could change the signature of `decode(r *http.Request, val interface{}, options DecoderOptions)` to just take an `io.Reader` instead of a whole request since it doesn't use any other parts of the request.

Reply:
but could not we provide a `decodeBytes` and use this one instead ? I really don't like modifying the request like that

Reply:
> but could not we provide a `decodeBytes` and use this one instead ? I really don't like modifying the request like that

That'd be just `json.Unmarshall` then I believe.

Thinking about Catalog usage, I'm not sure we still need the _do-it-all_ `DecodeWithJSONSchema`, and we should just use the `ParseJSONInput()` and `jsonschema.ValidateJSON()` functions we added and then just unmarshal into our struct since the `decode` logic is superfluous when we've already validated with a JSON schema.

Which is why I was thinking of depreciating `DecodeWithEmbeddedJSONSchema` and `DecodeWithJSONSchema`,  just keep exposing the `Decode()` for when we don't rely on JSON schemas and the `ParseJSONInput()` + `jsonschema.ValidateJSON()` for the JSON schema cases.

And we'd just compose with those primitives as needed in the application code instead of having a one-size fits all function.

Reply:
So the API we expose from `go-pkg` would then be:
```go
// web package

func Decode(r *http.Request, val interface{}) error {} // <- For when we don't need JSON Schema validation

func ParseJSONInput(request *http.Request) ([]byte, error) {} // <- For when we want to get the JSON byte slice to feed it to the Validate function


// jsonschema package

func ValidateJSON(jsonData []byte, jsonSchema []byte) error {}

```

Do you think it'd be a nicer API overall ?

Reply:
Otherwise, if we want to keep `DecodeWithJSONSchema`, I think we can forego the decoding part there and just unmarshal into the struct for those functions. I'll try to see if it breaks the current behaviour.